dnsapi/dns_edgedns.sh

```diff
@@ -0,0 +1,451 @@
+    _log "$(printf "Data: %s" "$data")"
+  fi
+
+  if [ "$_code" = "403" ]; then
+    _err "access denied make sure your Edgegrid cedentials are correct."
+  fi
+
+  echo "$_code"
+  return 1
+}
+
+_edgedns_eg_timestamp() {
+  _eg_timestamp=$(date -u "+%Y%m%dT%H:%M:%S+0000")
+  _debug3 "_eg_timestamp" "$_eg_timestamp"
+}
+
+_edgedns_new_nonce() {
+  _nonce=$(echo "EDGEDNS$(_time)" | _digest sha1 hex | cut -c 1-32)
+  _debug3 "_nonce" "$_nonce"
+}
+
+_edgedns_make_auth_header() {
+  _debug "Constructing Auth Header"
+  _edgedns_eg_timestamp
+  _edgedns_new_nonce
+  # "Unsigned authorization header: 'EG1-HMAC-SHA256 client_token=block;access_token=block;timestamp=20200806T14:16:33+0000;nonce=72cde72c-82d9-4721-9854-2ba057929d67;'"
+  _auth_header="$(printf "EG1-HMAC-SHA256 client_token=%s;access_token=%s;timestamp=%s;nonce=%s;" "$AKAMAI_CLIENT_TOKEN" "$AKAMAI_ACCESS_TOKEN" "$_eg_timestamp" "$_nonce")"
+  _secure_debug2 "Unsigned Auth Header: " "$_auth_header"
+
+  _edgedns_sign_request
+  _signed_auth_header="$(printf "%ssignature=%s" "$_auth_header" "$_signed_req")"
+  _secure_debug2 "Signed Auth Header: " "${_signed_auth_header}"
+}
+
+_edgedns_sign_request() {
+  _debug2 "Signing http request"
+  _edgedns_make_data_to_sign "$_auth_header"
+  _secure_debug2 "Returned signed data" "$_mdata"
+  _edgedns_make_signing_key "$_eg_timestamp"
+  _edgedns_base64_hmac_sha256 "$_mdata" "$_signing_key"
+  _signed_req="$_hmac_out"
+  _secure_debug2 "Signed Request" "$_signed_req"
+}
+
+_edgedns_make_signing_key() {
+  _debug2 "Creating sigining key"
+  ts=$1
+  _edgedns_base64_hmac_sha256 "$ts" "$AKAMAI_CLIENT_SECRET"
+  _signing_key="$_hmac_out"
+  _secure_debug2 "Signing Key" "$_signing_key"
+
+}
+
+_edgedns_make_data_to_sign() {
+  _debug2 "Processing data to sign"
+  hdr=$1
+  _secure_debug2 "hdr" "$hdr"
+  _edgedns_make_content_hash
+  path="$(echo "$_request_url_path" | tr -d "\n\r" | sed 's/https\?:\/\///')"
+  path="${path#*$AKAMAI_HOST}"
+  _debug "hier path" "$path"
+  # dont expose headers to sign so use MT string
+  _mdata="$(printf "%s\thttps\t%s\t%s\t%s\t%s\t%s" "$_request_method" "$AKAMAI_HOST" "$path" "" "$_hash" "$hdr")"
+  _secure_debug2 "Data to Sign" "$_mdata"
+}
+
+_edgedns_make_content_hash() {
+  _debug2 "Generating content hash"
+  _hash=""
+  _debug2 "Request method" "${_request_method}"
+  if [ "$_request_method" != "POST" ] || [ -z "$_request_body" ]; then
+    return 0
+  fi
+  _debug2 "Req body" "$_request_body"
+  _edgedns_base64_sha256 "$_request_body"
+  _hash="$_sha256_out"
+  _debug2 "Content hash" "$_hash"
+}
+
+_edgedns_base64_hmac_sha256() {
+  _debug2 "Generating hmac"
+  data=$1
+  key=$2
+  encoded_data="$(echo "$data" | iconv -t utf-8)"
+  encoded_key="$(echo "$key" | iconv -t utf-8)"
+  _secure_debug2 "encoded data" "$encoded_data"
+  _secure_debug2 "encoded key" "$encoded_key"
+
+  data_sig="$(echo "$encoded_data" | tr -d "\n\r" | ${ACME_OPENSSL_BIN:-openssl} dgst -sha256 -hmac "$encoded_key" -binary | _base64)"
```

Comment:
don't use `openssl`.  please use `_hmac()` function instead.

Reply:
if I use _hmac, our API server returns a signature error. Its very particular.  I need to leave as is.

Reply:
it's not reasonable to me. why?

Reply:
@Neilpang Akamai API Server - Client communication uses a very specific, algorithmic based verification process to validate the http exchange. The Server uses a specific sequence of algorithmic steps to calculate the signature and compare to the signature generated by the client and included in the header. The calculated signature includes the remainder of the header as well as the body.  Two of the three Auth secrets are included in the header. The third, known to the server and client, is used to generate the signature on each side for comparison.

This is the first shell script implementation of the client. I had to reverse engineer the signature generation steps using existing golang and python implementations as a guide. The signature generation has to be EXACTLY the same as the one used by the server. This sig generation logic is very finnicky and has long caused angst. Trust me, it took weeks to get right!

The logic in the _hmac function is not the same as the logic I've used; apparently different enough to throw off the signature comparisons. Specifically,

_hmac:${ACME_OPENSSL_BIN:-openssl} dgst -"$alg" -mac HMAC -macopt "hexkey:$secret_hex" -binary 2>/dev/null || ${ACME_OPENSSL_BIN:-openssl} dgst -"$alg" -hmac "$(printf "%s" "$secret_hex" | _h2b)" -binary

-- versus--

edgedns:  ${ACME_OPENSSL_BIN:-openssl} dgst -sha256 -hmac "$encoded_key" -binary 

If you are interested, you can see the other two reference client implementations at  https://github.com/akamai/AkamaiOPEN-edgegrid-golang/blob/master/edgegrid/signer.go and https://github.com/akamai/AkamaiOPEN-edgegrid-python/blob/master/akamai/edgegrid/edgegrid.py

Please approve my PR. Thank you.

Reply:
I believe you can use `_hmac`  function instead.

please try it more.

you can also give me an example of inputs and output hmac value, so that I can help you.